standard-covoiturage_openapi.yaml

```diff
@@ -526,6 +554,113 @@ components:
           items:
             $ref: '#/components/schemas/JourneySchedule'
 
+    CarpoolBookingEvent:
```

Comment:
What would be the advantages of splitting the booking between the `CarpoolBookingEvent` and the actual `CarpoolBooking`? Alternatively, since there's always one booking with the same id, we could upsert it every time there's a change.

Reply:
There is few reasons to that : 
- posting a `Booking` on an endpoint named `/booking_events` doesn't seem right / comply with REST best practises.
- upsert is very powerful (for the reasons explained before) but not well documented. Making the operation an event feed makes it clear and REST compatible.
- the "event" layer can embed all the necessary metadata that have nothing to do with the booking itself (authentication, idempotency, ...).

This is highly inspired by webhook APIs like the one of [Stripe API](https://stripe.com/docs/api/events/object) that have proven to be robust for years.

Reply:
Hi @pierrecamilleri,

Thank you for your feedbacks. 

> - The API name could as well be /bookings

It would collide with the "booking by API" feature (we don't want that). More than that, it would be totally unclear to post "events" on an endpoint called `/bookings`. And using "events" serve additional purpose (see below).

> The PUT method requests that the enclosed entity be stored under the supplied Request-URI.

The created booking would never be retrievable through the `/bookings` route. So this precise definition goes against using `PUT /bookings` for upserts. Generally speaking, upsert is not supported by the HTTP standard.

> - It does not feel weird to me to have a passengerId inside the Booking object. I am not sure about idempotency.

It is not that it would feel weird or not. It is that the booking "owner" (passenger or driver) identity should be inferred by authentication, not arbitrary set (for security and reliability purpose). 

The suggested design is highly inspired by battle tested "events" APIs that serve the exact same purpose (reflecting changes of data from one system to another). See [Stripe](https://stripe.com/docs/api/events/object), [Uber](https://developer.uber.com/docs/riders/guides/webhooks) or [Zendesk](https://developer.zendesk.com/api-reference/event-connectors/events-for-amazon-eventbridge/tickets/) for example. 

In all these APIs, the "subject" is wrapped inside an event structure.  This allows [a] a fully HTTP compatible implementation [b] clear separation between the subject data ("booking" for us) and metadata (ex: authentication, timestamp of the event, idempotency mechanisms, event type if tomorrow it serves a purpose, ...) [c] no collision / confusion with the subject API itself (ie: the booking API for us).

Reply:
Thanks, I think I see your point. `PUT /bookings` would suggest that the transport operator creates/updates a `Booking` object for later retrieval, while `POST /booking_event` makes it clear that information is provided about an event for processing by the MaaS platform (with no intention of retrieval or injunction on how to process the data).

Reply:
> The suggested design is highly inspired by battle tested "events" APIs that serve the exact same purpose (reflecting changes of data from one system to another). See Stripe, Uber or Zendesk for example.

I would propose to not take advise from big corporations with lots of engineers too seriously for our needs. They might as well not be compatible in terms of needs and resources.

I tend to agree with @pierrecamilleri 's first comment on this thread [here](https://github.com/fabmob/standard-covoiturage/pull/5#discussion_r870541576). I still fail to see why we need to "complexify" this endpoint with an event layer when a simple `/booking` one would suffice. If the name conflicts with another, perhaps we could consider discussing renaming it instead of adding an extra layer of abstraction?

Reply:
@riclage 

> I would propose to not take advise from big corporations with lots of engineers too seriously for our case. They may not be compatible in terms of needs and resources.

My only was to show this architecture was not an "exoticism" I was pushing, but widespread. Don't worry we have been using this pattern extensively at Klaxit (as emitter & consumer) for years, with a team of <= 5 backend engineers. And I can garanty that the added layer solve a lot of misunderstandings & bugs (so a lot of engineering time at the end).

Reply:
As discussed in a meeting, OK for me to go with the events proposal you put forward.